From 09ff311f4d10ba51a1ff301346fcbbf9d98b2705 Mon Sep 17 00:00:00 2001
From: OneFeiFan <3022504358@qq.com>
Date: Thu, 29 Oct 2026 01:52:56 +0800
Subject: [PATCH] blk-mq: map ctx and hctx before limit_depth
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Move the ctx/hctx mapping before the elevator limit_depth callback.
ADIOS uses limit_depth and needs data->ctx/data->hctx to be valid there.

Origin: torvalds/linux 6259151c04d4（block: Call .limit_depth() after .hctx has been set），按 5.10 的请求分配路径适配
Upstream-Status: Backport
KABI: none
Depends-On: none
Tested-On: yudi 5.10.272 (ADIOS boot test)
---
 block/blk-mq.c | 11 +++++------
 1 file changed, 5 insertions(+), 6 deletions(-)

diff --git a/block/blk-mq.c b/block/blk-mq.c
index 96ede8fa0..75908b931 100644
--- a/block/blk-mq.c
+++ b/block/blk-mq.c
@@ -365,6 +365,10 @@ static struct request *__blk_mq_alloc_request(struct blk_mq_alloc_data *data)
 	if (data->cmd_flags & REQ_NOWAIT)
 		data->flags |= BLK_MQ_REQ_NOWAIT;
 
+retry:
+	data->ctx = blk_mq_get_ctx(q);
+	data->hctx = blk_mq_map_queue(q, data->cmd_flags, data->ctx);
+
 	if (e) {
 		/*
 		 * Flush requests are special and go directly to the
@@ -375,12 +379,7 @@ static struct request *__blk_mq_alloc_request(struct blk_mq_alloc_data *data)
 		    e->type->ops.limit_depth &&
 		    !(data->flags & BLK_MQ_REQ_RESERVED))
 			e->type->ops.limit_depth(data->cmd_flags, data);
-	}
-
-retry:
-	data->ctx = blk_mq_get_ctx(q);
-	data->hctx = blk_mq_map_queue(q, data->cmd_flags, data->ctx);
-	if (!e)
+	} else
 		blk_mq_tag_busy(data->hctx);
 
 	/*
-- 
2.43.0

